### PR TITLE
ENH looser ABI compat for pybind11

### LINKED ISSUE #77: Does `pybind11-abi` shield from using different versions of GCC
I'm having problems due to different libraries having been compiled using different versions of GCC, see https://github.com/pybind/pybind11/issues/3751 .

My question: does having 
```
host:
- pybind11-abi
```
shield me from this in the future? 

Is there a way to detect such an issue also for a local build of library being used to a build on conda-forge ?

### PR DESCRIPTION
**Pull request by beckermr**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

closes #77

### CODE REVIEW

**Comment by beckermr** [comment]
@conda-forge-admin rerender

**Comment by beckermr** [comment]
@conda-forge/core @conda-forge/pybind11 @isuruf this is ready for review!

File: recipe/meta.yaml [hunk 1]
```diff
@@ -14,10 +14,11 @@ source:
   sha256: {{ sha256 }}
 
 build:
-  number: 0
+  number: 1
```
Comment (isuruf) [inline recipe/meta.yaml:17]: Build number bump is not enough. You need a version number bump

Comment (beckermr) [reply recipe/meta.yaml:17]: We can mark build 0 broken and I think we'll be fine.

Comment (beckermr) [reply recipe/meta.yaml:17]: We also need to mark any v4 pybind11 Abi packages as broken too.

Comment (isuruf) [reply recipe/meta.yaml:17]: Wouldn't it be better to bump version to v5 and rebuild?

Comment (beckermr) [reply recipe/meta.yaml:17]: The pybind11-abi version is supposed to track the pybind11 internals version. So we'd make them out of sync. We could append a minor version to it that is the conda-forge one. This has the risk of being an issue it pybind11 changes its scheme later. We could also preprend a 0. to it which avoids the upstream changes.

Comment (beckermr) [reply recipe/meta.yaml:17]: We could also just let conda-forge have an extra version increment.

Comment (beckermr) [reply recipe/meta.yaml:23]: See my latest commit for a possible fix.

Comment (isuruf) [reply recipe/meta.yaml:23]: Not a fan of that. Why not use a `epoch`?

Comment (beckermr) [reply recipe/meta.yaml:23]: We can go to epochs. They are ugly and I do not like them but I don't have strong opinions here.

Comment (beckermr) [reply recipe/meta.yaml:23]: Hopefully mamba can handle them?

Comment (beckermr) [reply recipe/meta.yaml:21]: OK. epochs pushed.

**Comment by beckermr** [comment]
@conda-forge-admin rerender

File: recipe/meta.yaml [hunk 2]
```diff
@@ -34,7 +34,7 @@ outputs:
         - include/pybind11/detail/internals.h
       commands:
         # make sure the internals version matches the package version
-        - if [ $(grep "#define PYBIND11_INTERNALS_VERSION" include/pybind11/detail/internals.h | cut -d' ' -f3) != "{{ abi_version }}" ]; then exit 1; fi
+        - if [ "$(grep \"#    define PYBIND11_INTERNALS_VERSION 4\" include/pybind11/detail/internals.h | cut -d' ' -f3)" != "{{ abi_version }}" ]; then exit 1; fi
```
Comment (beckermr) [inline recipe/meta.yaml:37]:
```suggestion
        - if [ "$(grep \"#    define PYBIND11_INTERNALS_VERSION 4\" include/pybind11/detail/internals.h | cut -d' ' -f4)" != "{{ abi_version }}" ]; then exit 1; fi
```

**Comment by beckermr** [comment]
@conda-forge-admin rerender

**Comment by beckermr** [comment]
@conda-forge-admin rerender

**Comment by beckermr** [comment]
@conda-forge-admin rerender

**Comment by beckermr** [comment]
@isuruf This is ready for another look!

**Comment by beckermr** [comment]
Any other comments here @isuruf?

**Comment by tdegeus** [comment]
Great! Will a guide be added to the knowledge base?

**Comment by tdegeus** [comment]
Any news on this?

**Comment by beckermr** [comment]
Needs to be merged locally and rerender. Was waiting on review from @isuruf plus migration work.

**Comment by tdegeus** [comment]
I'm experience quite some extra work from the ABI check (and the fact that I'm not on the same GCC version at our computing center). It would be great if this could be merged soon!

**Review by varlackc — COMMENTED** [review]
We are missing a `dev_url` in the `meta.yaml` file

  https://github.com/conda-forge/pybind11-feedstock/blob/4d763b81881c6d6cd6bb760f6836cfb68a173e4d/recipe/meta.yaml#L92&L99

  We might be able to use the following `dev_url`
  dev_url: https://github.com/pybind/pybind11/

**Comment by jakirkham** [comment]
@varlackc please raise that as a new issue

**Comment by isuruf** [comment]
> Let me add one correction: simply using different versions of GCC alone is not enough to cause issues -- pybind11 queries the ABI compatibility tag of libstdc++ (__GXX_ABI_VERSION). And that's where this whole thing gets kind of tricky: when pybind11 isolates different extensions, it's does so intentionally because a change in the ABI tag. So it is your word against those of the libstdc++ authors, who are saying that there was an ABI break (which you then want to override/disable manually.)

GXX_ABI_VERSION is too broad. Yes, there might have been an ABI break because of name mangling issues, but there is a compatibility scheme that adds a aliased symbol. See `-fabi-compat-version` in https://gcc.gnu.org/onlinedocs/gcc-13.2.0/gcc/C_002b_002b-Dialect-Options.html

**Comment by tdegeus** [comment]
I can comment on what causes problems in my case. Suppose that I have a library `foo` that provides `class Foo`  with a pybind11 API. When I then have a library `bar` with say a method `void bar(Foo& arg)` then things would just work from the Python side using `from foo import Foo`.  

In practice, this breaks when the Linux compiler on conda-forge is updated. Either of two solutions would be satisfactory *for me*:
1. Relax pybind11's strictness on the compatibility check.
2. Trigger rebuilds for the dependencies such that all packages with the same `pybind1_abi` are compatible.

**Comment by beckermr** [comment]
pybind11's ABI story has changed a lot. closing this per #96

**Pull request closed without merging**